aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,776 @@
+          } else {
+            values_row[k] = out_values_row[k] * (tmp_row[k]);
+          }
+        }
+      }
+    }
+    index += step;
+  }
+}
+
+std::tuple<
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>>
+get_pools(
+    const Tensor& indices,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+
+  thrust::host_vector<int64_t> host_strides(ndim, 1);
+
+  if (ndim > 1) {
+    for (int64_t i = ndim - 2; i >= 0; i--) {
+      host_strides[i] = host_strides[i + 1] * (i + 1 == dim ? 1 : sizes[i + 1]);
+    }
+  }
+  thrust::device_vector<int64_t> strides = host_strides;
+  int64_t* strides_ptr = strides.data().get();
+
+  auto indices_accessor = indices.packed_accessor<int64_t, 2>();
+
+  thrust::device_vector<int64_t> pool_indices(nnz);
+
+  thrust::transform(
+      thrust::make_counting_iterator(int64_t(0)),
+      thrust::make_counting_iterator(int64_t(nnz)),
+      pool_indices.begin(),
+      [indices_accessor, strides_ptr, dim, ndim] __device__(int64_t x) {
+        int64_t pool_index = 0;
+        for (int64_t j = 0; j < ndim; j++) {
+          if (j != dim) {
+            auto indices_row = indices_accessor[j];
+            auto stride = strides_ptr[j];
+            pool_index += stride * indices_row[x];
+          }
+        }
+        return pool_index;
+      });
+  int64_t* pools_ptr = pool_indices.data().get();
+
+  thrust::device_vector<int64_t> sorted_indices(nnz);
+  thrust::sequence(sorted_indices.begin(), sorted_indices.end(), 0);
+
+  thrust::sort(
+      sorted_indices.begin(),
+      sorted_indices.end(),
+      [pools_ptr] __device__(int64_t x, int64_t y) {
+        return pools_ptr[x] < pools_ptr[y];
+      });
+
+  thrust::device_vector<int64_t> count_set(nnz);
+
+  auto new_end = thrust::reduce_by_key(
+      sorted_indices.begin(),
+      sorted_indices.end(),
+      thrust::make_constant_iterator(int64_t(1)),
+      thrust::make_discard_iterator(),
+      count_set.begin(),
+      [pools_ptr] __device__(int64_t x, int64_t y) {
+        return pools_ptr[x] == pools_ptr[y];
+      });
+
+  auto new_sz = thrust::distance(count_set.begin(), new_end.second);
+
+  count_set.resize(new_sz);
+
+  auto pool_size = count_set.size();
+  thrust::device_vector<int64_t> device_pool_offsets = count_set;
+
+  thrust::exclusive_scan(
+      device_pool_offsets.begin(),
+      device_pool_offsets.end(),
+      device_pool_offsets.begin());
+
+  return std::make_tuple(sorted_indices, device_pool_offsets, count_set);
+}
+
+template <class scalar_t>
+std::tuple<
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<int64_t>,
+    thrust::device_vector<scalar_t*>>
+compute_pool_max(
+    const Tensor& indices,
+    const Tensor& values,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
```

Comment:
Please merge the template functions `get_pool` and `compute_pool_max` into a single template function that has an additional parameter, say `require_mx_rows`, about requesting the computation of `mx_rows`. This will reduce code replication in these functions considerably.